### PR TITLE
bpf: Added eBPF code coverage reporting

### PR DESCRIPTION
**Pull request by dylandreimerink**
This commit adds code coverage for eBPF tests using the new test
framework. The code coverage works based on CoverBee, a tool that was
developed for this purpose.

CoverBee instruments already compiled ELF objects without modifications
to the build chain. It constructs a control flow graph from the program.
Each node containing a block of instructions that all run together.
At the start of each of these blocks, instrumentation is added which
increments a counter in a map value at an index based on the block.
This map is then read after running the tests, BTF line info is then
used to apply the counts of each block to the source code using the
go cover library.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

### CODE REVIEW

File: test/bpf_tests/bpf_test.go [hunk 1]
```diff
@@ -33,9 +34,17 @@ import (
 	"github.com/cilium/cilium/pkg/byteorder"
 	"github.com/cilium/cilium/pkg/datapath/link"
 	"github.com/cilium/cilium/pkg/monitor"
+	"github.com/cilium/coverbee"
+	"github.com/dylandreimerink/gocovmerge"
+	"golang.org/x/tools/cover"
```
Comment (christarazi) [inline test/bpf_tests/bpf_test.go]: Nit: These should be in the second grouping as 3rd party imports

**Review by christarazi — APPROVED** [review]

**Review by rolinh — APPROVED** [review]

File: test/bpf_tests/bpf_test.go (hunk 1 above)
Comment (joamaki) [reply test/bpf_tests/bpf_test.go]: Tip: `goimports -local github.com/cilium/cilium` will sort the imports correctly. Consider adding it into your editor's file save hooks.

**Review by joamaki — APPROVED** [review]

File: test/bpf_tests/bpf_test.go [hunk 2]
```diff
@@ -30,12 +31,21 @@ import (
 	"github.com/vishvananda/netlink/nl"
 	"google.golang.org/protobuf/encoding/protowire"
 
+	"github.com/cilium/coverbee"
```
Comment (christarazi) [inline test/bpf_tests/bpf_test.go]: Sorry to keep nit picking, but these imports should be grouped together

Comment (dylandreimerink) [reply test/bpf_tests/bpf_test.go]: If course, will fix it. @joamaki `goimports -local github.com/cilium/cilium` doesn't fix everything :smile:

**Review by qmonnet — APPROVED** [review]
Nice!

**Comment by dylandreimerink** [comment]
The CI seems to be having a bad day, looks like the core issue is due to connection issues to quay. I think we can still safely merge since this PR is related to the testing framework and not agent or datapath code and the linters and such are all green.

**Comment by qmonnet** [comment]
@dylandreimerink the new coverage reports are supposed to be run by the CI, right? If that's the case, I'd like to have at least a run that shows that the updated test with CoverBee is passing, before we merge, please

**Comment by dylandreimerink** [comment]
Code coverage for eBPF is not yet enabled in the CI, currently only invoked when running `make -C test run_bpf_tests COVER=1`. We should at some point integrate it in the CI so we can track coverage over time, but this requires some work together with people how are more knowledgeable about our CI and where to put artifacts. 

The most important CI check for this code would be to actually run this changed go code. But the CI doesn't do this because no datapath code has been changed. Is there some way to trigger the 'build datapath' job manually?

File: go.mod [hunk 3]
```diff
@@ -29,6 +30,7 @@ require (
 	github.com/davecgh/go-spew v1.1.1
 	github.com/docker/docker v20.10.17+incompatible
 	github.com/docker/libnetwork v0.8.0-dev.2.0.20210525090646-64b7a4574d14
+	github.com/dylandreimerink/gocovmerge v1.0.1
```
Comment (pchaigno) [inline go.mod:33]: Any chance you could submit your patches upstream to avoid relying on a fork in the future?

File: test/bpf_tests/bpf_test.go [hunk 4]
```diff
@@ -56,26 +80,60 @@ func TestBPF(t *testing.T) {
...
+	var (
+		coll *ebpf.Collection
+		cfg  []*coverbee.BasicBlock
+		err  error
+	)
+
+	opts := ebpf.CollectionOptions{
 		Programs: ebpf.ProgramOptions{
-			LogSize: 4 << 20, // 4 MiB
+			LogSize: 32 << 20, // 32 MiB, not needed in most cases, but especially when running instrumented code.
```
Comment (pchaigno) [inline test/bpf_tests/bpf_test.go]: Did you mean this?
```suggestion
			LogSize: 32 << 20, // 32 MiB, not needed in most cases, except when running instrumented code.
```

**Review by pchaigno — APPROVED** [review]
LGTM, couple minor comments below.

**Comment by YutaroHayakawa** [comment]
@dylandreimerink Cool stuff! I tried this on my local, but the result seems odd. Below, I have a screenshot of the result of `conntrack.h`. Most of the lines are marked as not tracked and covered and uncovered lines are mixing up without branch. Do you have any idea why this happens?

スクリーンショット 2022-07-13 10 22 21

**Comment by YutaroHayakawa** [comment]
Also, when I turn on the instrumentation log, I hit the verifier log size limit. Maybe better to bump up?

```
$ make -C test run_bpf_tests COVER=1 INSTRLOG=1                                                                          /home/yutaro/WorkSpace/cilium
make: Entering directory '/home/yutaro/WorkSpace/cilium/test'
make -C ../bpf/tests all
make[1]: Entering directory '/home/yutaro/WorkSpace/cilium/bpf/tests'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/yutaro/WorkSpace/cilium/bpf/tests'
go test bpf_tests/*.go -exec sudo -bpf-test-path /home/yutaro/WorkSpace/cilium/bpf/tests  -coverage-report /home/yutaro/WorkSpace/cilium/bpf-coverage.html -instrumentation-log /home/yutaro/WorkSpace/cilium/test/bpf-instrumentation.log
--- FAIL: TestBPF (5.96s)
    bpf_test.go:136: Verifier logs larger than 32MiB, increase the log size passed to ebpf.NewCollectionWithOptions or decrease the size/complexity of the program
FAIL
FAIL	command-line-arguments	5.983s
FAIL
make: *** [Makefile:84: run_bpf_tests] Error 1
make: Leaving directory '/home/yutaro/WorkSpace/cilium/test'
```

**Comment by dylandreimerink** [comment]
> Most of the lines are marked as not tracked and covered and uncovered lines are mixing up without branch. Do you have any idea why this happens?

This is the result of highly optimized code. All of the untracked code(variable declarations and code within the `#ifdef`) simply didn't make it into the compiled binary, so coverbee can't track it. I am working on an coverbee update that improves this by combining the results with static analysis of the source code, but that is still going to take some time.

What you will also notice is that the `ct_state->rev_nat_index = entry->rev_nat_index;`, `ct_state->loopback = entry->lb_loopback;` and `ct_state->node_port = entry->node_port;` lines all have different amounts of coverage, while they should be executed together. I suspect this is also because of compiler optimization. Since `__ct_lookup` is inlined, the compiler likely checks which variables of the result are used and only includes the set statements for those fields.

>  I hit the verifier log size limit. Maybe better to bump up

Oh, hmm. If we are still hitting the limit at 32MiB, we should make it a bit smarter, I have some ideas for that.

**Comment by dylandreimerink** [comment]
> Oh, hmm. If we are still hitting the limit at 32MiB, we should make it a bit smarter, I have some ideas for that.

I have been playing around with automatically resizing the log. However due to the way it is currently implemented, coverbee will redo all of the instrumentation work every time we resize, which makes the process to slow. So I have just doubled the log size again, for now. I will have to restructure coverbee and the way we invoke it before we can make a more robus retry system.

File: go.mod (hunk 3 above)
Comment (dylandreimerink) [reply go.mod:33]: To be fair, the amount of code contained within the library is so small that I am just going to embed it directly into coverbee with a copyright notice. I will do that in a next coverbee update.

**Comment by dylandreimerink** [comment]
This PR doesn't touch any code that could cause the failures we are seeing in the CI, so it looks like those are flakes

**Pull request merged**